Create and renew teacher subscriptions using the Stripe Subscription API

brendanshean opened the pull request:
## WHAT
Move the handling of Teacher Subscriptions from server to the the Stripe Subscriptions API. 

NB: A subsequent PR will handle the transitioning of existing Subscriptions to renewal via Stripe.

## WHY
We'd like to streamline how we manage subscriptions

## HOW
There are three places where teachers have the ability to purchase subscriptions
1) At `/premium` there is a 'Buy Now' button
![Screen Shot 2022-04-06 at 2 36 14 PM](https://user-images.githubusercontent.com/2057805/162044939-2d6dbac9-c191-4c58-8a5b-51973b8d598b.png)

2) If a user is in trial mode, `/premium` will also show 'Upgrade to Premium Now'
![Screen Shot 2022-04-06 at 2 36 18 PM](https://user-images.githubusercontent.com/2057805/162044952-35e410a8-56fa-47f7-a2ce-5f33bca48476.png)

3) If a user had a subscription but it has now expired, a 'Renew Subscription' button appears on `/subscriptions` page.
![renew-subscription-via-ui](https://user-images.githubusercontent.com/2057805/162037519-f8cf406e-6076-4f5c-838d-559a1de1fabf.png)

Each of these buttons has a new jsx component `StripeCheckoutSessionButton` which when clicked, triggers a [post](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/client/app/bundles/Teacher/components/shared/StripeCheckoutSessionButton.jsx#L28) request to `StripeIntegration::CheckoutSessionsController`.  This controller [requests](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/controllers/stripe_integration/checkout_sessions_controller.rb#L8) from Stripe a Checkout::Session object which has a url that that the frontend will then [redirect](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/client/app/bundles/Teacher/components/shared/StripeCheckoutSessionButton.jsx#L39) to:  


![stripe-checkout-flow](https://user-images.githubusercontent.com/2057805/162038560-7892d737-42fb-48b6-984e-b747c9453092.png)

After the user successfully fills in the credit card info and clicks 'Subscribe', Stripe redirects to the `success_url` specified [here](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/controllers/stripe_integration/checkout_sessions_controller.rb#L49) at which point Stripe immediately sends a bunch of webhook events, which I'll say more about later.

The browser redirects to the `subscriptions/?checkout_session_id={CHECKOUT_SESSION_ID}` where `CHECKOUT_SESSION_ID`  is used to  [determine](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/controllers/subscriptions_controller.rb#L60) that the subscription was purchased successfully.  This boolean is used to determine if the `PremiumConfirmationModal` is show to the user after completion of purchase.
![Screen Shot 2022-04-06 at 2 17 41 PM](https://user-images.githubusercontent.com/2057805/162041811-99e3081d-2b21-4e11-ba34-97cb7961f254.png)

From there the the "Payment Method on File" should list the credit card along with the last4, which has to be calculated [differently](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/models/stripe_integration/subscription.rb#L5) since the previous [approach](https://github.com/empirical-org/Empirical-Core/blob/5cc36287f1ed2bdc3d7b2bedf5c04778b9e43a7a/services/QuillLMS/app/models/user.rb#L307) no longer works.

If the user clicks 'Edit Credit Card', a similar [flow](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/controllers/stripe_integration/billing_portal_sessions_controller.rb#L6) to the CheckoutSession one occurs:

![edit-card](https://user-images.githubusercontent.com/2057805/162043298-94558657-d155-42fd-85be-4103fdfbbc4b.png)

#### Webhooks
From a high level, Stripe sends various messages our endpoint `/stripe_integration/webhooks`.  From there our controller must respond to the events [immediately](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/controllers/stripe_integration/webhooks_controller.rb#L8) returning a 200 or 400 depending on the circumstances and kick off a background job `StripeIntegration::Webhooks::HandleEventWorker` to process the event.  The `EventHandlerFactory` determines which handler to use.  Currently there are:

`InvoicePaidEventHandler`
When [triggered](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/services/stripe_integration/webhooks/invoice_paid_event_handler.rb#L13) provisions a new Subscription using the `SubscriptionCreator`  [service object](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/services/stripe_integration/webhooks/subscription_creator.rb#L5)

NB.  In my Stripe RFC Schema, I had initially added a column `stripe_subscription_id` to the `Subscription` table.  However, Stripe persists that id across renewals, whereas our system creates a new subscription object each renewal.  To better reflect this modeling, I've instead added a `stripe_invoice_id` column which has a one-to-one correspondence with our `Subscription` table.  It also allows us to pull the relevant information from Stripe when provisioning our new Subscription objects.


`IgnoredEventHandler`
This is the current [list](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/services/stripe_integration/webhooks/ignored_event_handler.rb#L6) of events that I have observed (and at this point am ignoring).

`UnknownEventHandler`
All other events are considered unknown and will [log](https://github.com/empirical-org/Empirical-Core/blob/f3500bd340e0e922c6a914d436dafa00472feb01/services/QuillLMS/app/services/stripe_integration/webhooks/unknown_event_handler.rb#L9) an error so that we can then either update our codebase with a new Handler or add that unknown event to the ignored list.

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
https://www.notion.so/quill/Implement-a-new-direct-subscription-process-that-utilizes-Stripe-subscriptions-229f2cba7d5e4dbf8bf95388f1ab4209

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | YES
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | YES

brendanshean committed:
Add PlanSerializer

brendanshean committed:
Add backend code for stripe checkout session

brendanshean committed:
Add front-end code for checkout session

brendanshean committed:
Add backend code to handle checkout.session.completed webhook

brendanshean committed:
wip

brendanshean committed:
enable some other stripe checkout buttons

brendanshean committed:
add stripe_invoice_id to subscription model

brendanshean committed:
Add ignored stripe webbhook event handler

brendanshean committed:
Load modal after completion of invoice.paid

brendanshean committed:
Update change credit card and last_four

brendanshean committed:
Cleanup dead code

brendanshean committed:
Fix last4

brendanshean committed:
Add simple delegator for stripe subscription

brendanshean committed:
Fix failing specs

brendanshean committed:
Fix snapshots

brendanshean committed:
Remove select credit card modal

brendanshean committed:
Misc

emilia-friedberg reviewed (approved):
Thank you for this incredibly clear and thorough explainer - it definitely made the code much easier to follow. One question I have is around the `CHECKOUT_SUCCESS_URL`/banner logic - if the webhooks don't complete until after the redirect is completed, will the user just never see the banner? Maybe it's generally fast enough that we don't need to worry, but we might want to consider using Pusher or something when the background job completes if it has not yet succeeded.

brendanshean commented:
@emilia-friedberg That's a very astute observation.  I kind of settled with the hack [here](https://github.com/empirical-org/Empirical-Core/blob/111926fcd5578336362b2d363d073d7db50cc1c6/services/QuillLMS/app/models/subscription.rb#L350), but Pusher seems like a better solution.  Thank you for the suggestion.

emilia-friedberg commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -616,6 +603,10 @@ def completed_activity_count
   end
 
+  def subscription_status
+    subscription&.subscription_status || last_expired_subscription&.subscription_status
```
I might be missing something, but is it a problem that the old version of code does return an object if the user has no subscription, and here we'd always be returning nil?

brendanshean replied:
This was tricky, so definitely worth discussing.  I extracted subscription_status from the [controller](https://github.com/empirical-org/Empirical-Core/pull/9005/files#diff-aa578c282750da885731bb9b3b1d6f9a2a74028b046729e15c1c8e18acff5c5aL72) and moved it to the User model since it's asking a question of the user.  The original method has three branches for what the user has: 
1) a current subscription
2) an expired subscription
3) no subscriptions

For case 3), the original code had an explicit return, but the new disjunction `subscription&.subscription_status || last_expired_subscription&.subscription_status` should now cover case 3) via the safe_navigation operator `&.`

emilia-friedberg commented on services/QuillLMS/app/models/subscription.rb:
```diff
@@ -352,4 +342,33 @@ def self.create_with_school_or_user_join school_or_user_id, type, attributes
+  def self.stripe_purchase_completed?(checkout_session_id)
+    return false if checkout_session_id.nil?
+
+    # TODO: remove this blocking waits for webhooks to complete
+    sleep ENV.fetch('STRIPE_WAIT_FOR_WEBHOOK', 0).to_i
```
Oh, I see, this is handling my concern about the webhooks completing after page load?

brendanshean replied:
Definitely a hack 😜.  I think pusher will be a more sensible approach.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Teacher/components/subscriptions/SubscriptionStatus.jsx:
```diff
@@ -78,7 +86,7 @@ const SubscriptionStatus = ({ subscriptionType, showPurchaseModal, subscriptionS
     case DISTRICT_PREMIUM:
     case SCHOOL_PREMIUM_SCHOLARSHIP:
-      content.pCopy = schoolPremiumCopy(subscriptionType);
+      content.premiumCopy = schoolPremiumCopy(subscriptionType);
```
lol this is definitely an improvement

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Teacher/components/subscriptions/current_subscription.jsx:
```diff
@@ -151,13 +148,30 @@ export default class CurrentSubscription extends React.Component {
+
+    if ( !stripe_customer_id ) { return }
+
+    request.post({
```
I don't think you'd have much way of knowing this, but we've been trying to use a centralized `request` module that @anathomical made a while back rather than invoking `request` directly (by importing from here: https://github.com/empirical-org/Empirical-Core/blob/develop/services/QuillLMS/client/app/modules/request/index.js#L48). This is definitely something we could change as a follow-up card.

brendanshean replied:
Thank you for drawing attention.  I did notice the `request` is deprecated so I will update this now.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Teacher/components/shared/StripeCheckoutSessionButton.jsx:
```diff
@@ -0,0 +1,56 @@
+        and we will renew it automatically when your subscription ends.`
+      )
+    } else {
+      request.post({
```
See comment elsewhere about `requestPost`.

emilia-friedberg reviewed (approved):
Submitted last comment too early! A few more notes for your consideration. I'm really impressed by this effort.

brendanshean committed:
Update request with standardized requestPost

brendanshean committed:
Prework for pusher integration with stripe subscription purchase confirmation

dandrabik commented on services/QuillLMS/app/controllers/stripe_integration/checkout_sessions_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    private def success_url
+      "#{ENV['DEFAULT_URL']}/subscriptions?checkout_session_id={CHECKOUT_SESSION_ID}"
```
The `checkout_session_id` is a constant? Just flagging in case that was a mistake.

brendanshean replied:
I was confused by this in the [docs](https://stripe.com/docs/payments/checkout/custom-success-page#modify-success-url), but it's apparently a "template variable", allowing Stripe to populate it later.

dandrabik commented on services/QuillLMS/app/controllers/stripe_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,37 @@
+    rescue ActiveRecord::RecordNotUnique
+      head 200
+    rescue => e
+      NewRelic::Agent.notice_error(e)
```
A few days ago, I added a `ErrorNotifier.report(error)`. I'd like to start using that in other places (we log them in both Sentry and NR for now).
```suggestion
      ErrorNotifier.report(e)
```

brendanshean replied:
Cool!  I will update.

dandrabik commented on services/QuillLMS/app/controllers/stripe_integration/webhooks_controller.rb:
```diff
@@ -0,0 +1,37 @@
+      head 400
+    end
+
+    private def endpoint_secret
```
Would this be better as a constant, e.g. `ENDPOINT_SECRET = ENV.fetch('STRIPE_ENDPOINT_SECRET', '')`

dandrabik commented on services/QuillLMS/app/models/subscription.rb:
```diff
@@ -80,15 +83,26 @@ class Subscription < ApplicationRecord
+
   ALL_TYPES = OFFICIAL_FREE_TYPES.dup.concat(OFFICIAL_PAID_TYPES)
 
+  validates :stripe_invoice_id, allow_blank: true, format: { with: /\Ain_[0-9a-zA-Z]*\z/ }
```
Do we have documentation that this is a set format for `invoice.id`? Will it be loud if this fails (if they add an underscore or something).

brendanshean replied:
@dandrabik I couldn't find any documentation concerning the exact form, but all of the different models in the Stripe documentation followed this same format.  I've added a [validator](https://github.com/empirical-org/Empirical-Core/blob/da3ea2f9f97db42f1a9939fecb00a6553368666a/services/QuillLMS/app/models/validators/stripe_uid_validator.rb) to group the pattern and give more informative error reporting when validation fails (e.g. subscription [creation](https://github.com/empirical-org/Empirical-Core/blob/da3ea2f9f97db42f1a9939fecb00a6553368666a/services/QuillLMS/app/services/stripe_integration/webhooks/subscription_creator.rb#L72))

dandrabik commented on services/QuillLMS/app/models/subscription.rb:
```diff
@@ -80,15 +83,26 @@ class Subscription < ApplicationRecord
   end
 
+  def expired?
+    expiration <= Date.today
```
We might want to add this cop and move away from `Date.today` (in favor of Date.current or `Time.now.utc.to_date`), just to ensure the timezone is utc.

https://www.rubydoc.info/gems/rubocop/0.41.2/RuboCop/Cop/Rails/Date

brendanshean replied:
I will update this instance and add then add the Rubocop setting in a future PR since it will require a new gem: rubocop-rails.

dandrabik commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -243,13 +243,7 @@ def subscription_authority_level(subscription_id)
-      # otherwise they are good for purchase
-      true
-    end
+    !subscription || Subscription::TRIAL_TYPES.include?(subscription.account_type)
```
I think(?) this is just a `nil` check, so consider `subscription.nil?`for readability?

dandrabik commented on services/QuillLMS/app/serializers/plan_serializer.rb:
```diff
@@ -0,0 +1,29 @@
+  attributes :display_name, :price_in_dollars, :stripe_price_id
+
+  def price_in_dollars
+    object.price / 100
```
Price is an integer, let's make the denominator a float to avoid weird integer division.
```suggestion
    object.price / 100.0
```

<img width="350" alt="Screen Shot 2022-04-07 at 5 15 44 PM" src="https://user-images.githubusercontent.com/1304933/162314386-ac295a0d-b1eb-4870-823f-098af64ad964.png">

dandrabik commented on services/QuillLMS/app/services/stripe_integration/webhooks/subscription_creator.rb:
```diff
@@ -0,0 +1,94 @@
+      end
+
+      private def stripe_price_id
+        stripe_subscription&.items&.data&.first&.price&.id
```
It's nil checks all the way down! 🤣

dandrabik commented on services/QuillLMS/client/app/bundles/Teacher/components/scorebook/premium_banners/free_trial_status.jsx:
```diff
@@ -1,27 +1,18 @@
-        <button className='btn-orange' type='button'>Upgrade to Premium Now</button>
-      </a>
-    )
+  const premiumButton = (originPage == 'premium')
```
Copied code, but I think we want a threequal.
```suggestion
  const premiumButton = (originPage === 'premium')
```

dandrabik commented on services/QuillLMS/client/app/bundles/Teacher/components/subscriptions/current_subscription.jsx:
```diff
@@ -35,18 +35,15 @@ export default class CurrentSubscription extends React.Component {
-    const { lastFour, } = this.state
 
-    if (subscriptionStatus && subscriptionStatus.payment_method === 'Credit Card' && lastFour && authorityLevel) {
+    if (subscriptionStatus && subscriptionStatus.payment_method === 'Credit Card' && authorityLevel) {
```
This isn't your code, so feel free to punt for now, but we should consider using more constants for strings that are used for logic in the payments code, e.g. `'Credit Card'`.

brendanshean replied:
I would agree.  I'm planning on addressing this in one of the future stripe PRs.

dandrabik commented on services/QuillLMS/client/app/bundles/Teacher/components/subscriptions/current_subscription.jsx:
```diff
@@ -197,15 +234,21 @@ export default class CurrentSubscription extends React.Component {
   nextPlanAlertOrButtons(condition, renewDate) {
-    const { lastFour, } = this.state
     const { authorityLevel, subscriptionStatus, } = this.props
+    const { last_four } = subscriptionStatus
```
I'm noticing more mixed variable naming (some `camelcase`, some `snakecase`). Maybe this is coming from Stripe, so we're not in control of it, but it might be worth discussing with @emilia-friedberg and @eadams17 how we should determine which to use, so that if someone else is working on this code, it's clear what to use (or it can lead to errors/bugs down the road).

brendanshean replied:
I spoke with Emilia and since the subscriptionStatus is older code, we'll leave as-is for now.

brendanshean replied:
Long term I think the subscriptionStatus should be refactored and we can address it then.

dandrabik commented on services/QuillLMS/app/controllers/stripe_integration/billing_portal_sessions_controller.rb:
```diff
@@ -0,0 +1,22 @@
+    end
+
+    private def return_url
+      "#{ENV['DEFAULT_URL']}/subscriptions"
```
Is there a route_url we can use? Trying to get away from these manual urls.

dandrabik reviewed (approved):
Overall, looks good to me! Thanks for putting this together, it looks like a lot of work! Thanks for being so thorough! 

I made a bunch of small comments, but none of them are blockers (I think all are take or leave).

👍

brendanshean committed:
Fix purchase confirmation with Pusher

brendanshean committed:
Dan suggestions

brendanshean committed:
Fix failing spec

brendanshean commented:
@emilia-friedberg Just a heads-up, I'm re-requesting review due to the Pusher integration that wasn't in the original PR
It's basically these two commits:
pre-work: https://github.com/empirical-org/Empirical-Core/pull/9005/commits/a67fba8b8206a5d3d1d9a83e6211c3dc33de2af8
implementation:  https://github.com/empirical-org/Empirical-Core/pull/9005/commits/c6227750c3a638dfc9e058e175ad3ec809d602b8

brendanshean committed:
Fix return_url

emilia-friedberg reviewed (commented):
This looks great from the frontend but I'm not seeing the part where the backend is actually implementing the pusher notification? Maybe I've just got Monday morning brain.

brendanshean commented:
@emilia-friedberg My mistake, I forgot to include the third commit:  https://github.com/empirical-org/Empirical-Core/pull/9005/commits/c6227750c3a638dfc9e058e175ad3ec809d602b8

brendanshean commented:
@emilia-friedberg You're right.  It's not in there.  I guess it must have been succeeding the first time everytime.  I need to add in the code for `quill_retrieval_processing`

brendanshean committed:
Add in pusher trigger

brendanshean commented:
@emilia-friedberg I forgot that I removed it from the background job due to the race condition we talked about earlier.  I've added it back in now since we still want it to trigger if it's slower than the redirect:  https://github.com/empirical-org/Empirical-Core/pull/9005/commits/c47acdceee5d8cf888b134ec9cc9bb032fbe8e07

brendanshean committed:
Add some specs

brendanshean committed:
Fix pusher integration backend code

brendanshean commented:
@emilia-friedberg Here is the commit for the pusher integration backend: https://github.com/empirical-org/Empirical-Core/pull/9005/commits/aad026e30b21bb644ae631b5f06b756bda697f8f

I verified it working by temporarily adding a `sleep 15` in the worker job.

brendanshean committed:
Fix specs

cissyxyu commented on services/QuillLMS/app/controllers/pages_controller.rb:
```diff
@@ -439,8 +436,7 @@ def premium
 
-  def tutorials
-  end
+  def tutorials; end
```
Might we just want to clean up these empty endpoints? I could be wrong but I don't think we're intending to use them for anything.

brendanshean replied:
It seems to be [pointing](https://github.com/empirical-org/Empirical-Core/blob/develop/services/QuillLMS/app/views/pages/tutorials.html.erb) to a 'TutorialsApp' which has some [components](https://github.com/empirical-org/Empirical-Core/tree/develop/services/QuillLMS/client/app/bundles/Teacher/components/tutorials)

cissyxyu commented on services/QuillLMS/app/controllers/subscriptions_controller.rb:
```diff
@@ -56,40 +63,22 @@ def destroy
     @subscriptions = current_user.subscriptions
     @premium_credits = current_user.credit_transactions.map {|x| x.serializable_hash(methods: :action)}.compact
-    subscription_status
+    @stripe_invoice_id = StripeIntegration::StripeInvoiceIdFinder.run(checkout_session_id)
```
Is this supposed to be `StripeIntegration::StripeInvoiceIdFinder.new(checkout_session_id).run`? The class definition says that the initialization takes an argument but `run` does not.

brendanshean replied:
That particular service object inherits from [ApplicationService](https://github.com/empirical-org/Empirical-Core/blob/develop/services/QuillLMS/app/services/application_service.rb#L5) which is  syntactic sugar that allows to collapse `ServiceObject.new(*args, &block).run` into just `ServiceObject.run(*args, block)`

cissyxyu commented on services/QuillLMS/app/models/subscription.rb:
```diff
@@ -80,15 +83,26 @@ class Subscription < ApplicationRecord
     account_type && TRIAL_TYPES.include?(account_type)
   end
 
+  def expired?
```
I think we're using `expiration <= Date.today` in some other places in the code to check if a subscription is expired. Might be worth double checking to make sure we're using this new method in all those old places.

brendanshean replied:
That's a good point.  Dan made a [suggestion](https://github.com/empirical-org/Empirical-Core/pull/9005#discussion_r845563873) to change this so I did but I will probably revert to the original and then change all of them when adding in the Rubocop configuration.

cissyxyu commented on services/QuillLMS/app/models/subscription.rb:
```diff
@@ -147,7 +162,6 @@ def credit_user_and_de_activate
 
-
   def self.expired_today_or_previously_and_recurring
     Subscription.where('expiration <= ? AND recurring IS TRUE AND de_activated_date IS NULL', Date.today)
```
see above comment (we can use the new method `expired?` here)

cissyxyu commented on services/QuillLMS/app/models/subscription.rb:
```diff
@@ -218,16 +207,23 @@ def update_if_charge_succeeds
     end
+  end
+
+  def save_if_charge_succeeds(premium_type, school=nil)
```
Looks like we're changing this method to only save school premium subs. Might want to change the method name as well to reflect that so it's not confusing.

brendanshean replied:
I will most likely be removing this method when integrating the school subscription stuff.  I decided to leave it for now so to keep the commit diff down.

brendanshean committed:
Revert Date.today for now

emilia-friedberg reviewed (approved):
Pusher stuff looks good so everything LGTM! Cheers!

brendanshean committed:
More Dan suggestions

brendanshean commented:
@dandrabik There is one more [commit](https://github.com/empirical-org/Empirical-Core/pull/9005/commits/a5d9e1ddcb79ef79243247a3fd539bbccf61beac) based on your QA testing:
